replica/database.cc

```diff
@@ -2583,8 +2583,13 @@ const sstring& database::get_snitch_name() const {
 }
 
 dht::token_range_vector database::get_keyspace_local_ranges(sstring ks) {
+    const auto& keyspace = find_keyspace(ks);
+    if (keyspace.get_replication_strategy().is_per_table()) {
+        // return an empty range if each tables have their own effective_replication_map
+        return dht::token_range_vector();
+    }
```

Comment:
we should encode the inability of retrieving keyspace's local ranges with optional. empty meaning not available is dangerous. 

maybe we can have two interfaces, 

dht::token_range_vector database::get_keyspace_local_ranges(sstring ks);

that will throw, as today, if replication strategy is not vnode based.

and 

std::optional<dht::token_range_vector> database::maybe_get_keyspace_local_ranges(sstring ks);

which returns nullopt if replication strategy is per table.

then we patch sstable upgrade to use the one above.

Reply:
Yes, as-is this is dangerous. I think we should change this method to throw an exception and adjust callers to not call it if the keyspace uses tablets.

Reply:
@raphaelsc I have updated the code like you have suggested. Is this enough or should we go all the way and make `owned_ranges_ptr` optional for the entire call chain?

Reply:
@denesb `get_keyspace_local_ranges` in its current implementation  internally calls `get_vnode_effective_replication_map` and it already throws an exception for keyspaces using tablets.

Reply:
I think this new version is good. @denesb a good reason for maybe_get_keyspace_local_ranges() is that compaction layer doesn't have to be aware of replication layer. as far as it's concerned, it only wants to know if per-keyspace ranges is available on behalf of a table. The series look good to me.

Reply:
Right, argree.